### ISSUE TITLE
Integrate my launch.json plugin

### ISSUE THREAD

**Issue by nbfalcon** [emacs-lsp/dap-mode#324]
launch.json is the de-facto standard for debugger launch configurations, and as such dap-mode should support that.
I have written a plugin (https://github.com/nbfalcon/dap-mode-launch-json) to add that feature, and would love to have it merged!

Feedback is welcome, and I would be happy to provide assistance if needed.

**Comment by yyoncho** [comment]
This is something definitely on the roadmap of dap-mode. I will try to provide a feedback for the plugin and how do I see the integration with dap-mode. You may turn it into a PR to make leaving the feedback easier to track.

**Comment by ericdallo** [comment]
Nice!
I'd like to see a PR adding this feature too :)

**Comment by nbfalcon** [comment]
My current concerns are the following:
- I define functions named projectile-*, which I should probably try to get into mainline projectile. Here I have a question: does dap-mode depend on some specific minimum version of projectile? If so, then how should we handle this?
- I define various functions named buffer-*; these are buffer filename related utilities, and I am unsure where I should try to merge them into

**Comment by ericdallo** [comment]
@nbfalcon it's not good to have functions in you lib that doesn't has you lib's name prefix, to avoid other libs with same function name, so you may rename it adding your lib prefix name.
If this comes to `dap-mode`, we may have the prefix `dap-buffer-*` and etc

**Comment by nbfalcon** [comment]
Ok, I will do that. But I want to get the projectile-* into mainline projectile, and I am currently writing the PR for that. However, that would mean that, if my plugin were merged, dap-mode would have to depend on a very new version of projectile. Would there be any problems with that?

**Comment by ericdallo** [comment]
I don't think so, it's pretty common we merge fixes and features on other projects and lsp-mode use it, it's not the best approach, but we recommend the user to update the package to the latest if face any problem.
What we can do to is check if the function is defined(if not, then the projectile is outdated) and not offer the feature, despite that this looks a major feature for dap-mode, so I'm not sure we should do that

**Comment by yyoncho** [comment]
dap-mode should not depend on projectile. There is lsp-workspace-root function which you should use for finding the project root.

**Comment by yyoncho** [comment]
The rest of `projectile-*` in your package functions are present in `f` package.

**Comment by nbfalcon** [comment]
Ok. I will update my plugin to not use projectile.

**Comment by nbfalcon** [comment]
Can we depend on f?

**Comment by yyoncho** [comment]
We already depend on `f`.

**Comment by nbfalcon** [comment]
What would be the f equivalent for projectile-relative-dirname as defined in my plugin?

**Comment by yyoncho** [comment]
`f-relative`

**Comment by nbfalcon** [comment]
I have cleaned up my plugin: all functions now have a dap- prefix; there was a major architectural bug related to with-temp-buffer that I had to fix, and my plugin now works correctly. The are no flycheck warnings and all functions are documented. You can now look at merging it into mainline, as I think my plugin is ready.

As for pull requests: I am unsure how to proceed here. The options I can see are adding my plugin as a separate file (as it exists now) or merging the file's contents into dap-mode.el.

The architecture of my plugin is as follows:
It contains a variable that I intend to be user-visible (and it is well documented): dap-launch-json-variables. It stores the regexes of all variable names (between ${ and }) with a callback, value, or symbol that is evaluated as a replacement (the value is expanded with dap--expand-poly-type, see its docstring for details).

The plugin has a main entry point: dap-debug-launch-json. It reads the project's launch.json file (the project is found with lsp-workspace-root, the launch.json must be at the root of the project) and presents the user with a list of launch configurations stored therein that they may choose. The launch configurations are found in the {"configurations" : ...} block. If it does not exist, the root json object is used as the launch configuration, as is the documented behavior of VSCode.

All launch configurations are plists as can be passed to dap-debug directly.

The variable expansion code is not launch.json specific, and it could be used directly with dap-mode launch.json plists. To expand the variables in a regular dap-mode launch configuration the function dap--launch-json-expand-variables can be used.

Perhaps the launch.json launch configurations should be shown in dap-debug directly?

**Comment by ericdallo** [comment]
@nbfalcon Nice! thanks for the hard work, I think you can open a PR adding as another file `dap-launch.el` then we can review the code and help with suggestions, what do you think?

**Comment by nbfalcon** [comment]
Will do.

**Comment by nbfalcon** [comment]
PR #325

**Comment by shackra** [comment]
uh, any chance of getting the usage of this feature documented?

**Comment by nbfalcon** [comment]
Im working on it. I'll tell you here meanwhile, since there isn't much to it:
1. Create a launch.json file in your project root (just like in VSCode; all variables that make sense in Emacs are supported!)
2. Run dap-debug
3. All the launch configurations from your launch.json file should now show up at the end
Happy debugging :)

**Comment by ericdallo** [comment]
Let me know if you need some help with the documentation @nbfalcon, You can open a PR adding a new "Launch configuration" page describing all these, first create a link to the new page [here](https://github.com/emacs-lsp/dap-mode/blob/master/mkdocs.yml#L4-L17) then create the markdown page [here](https://github.com/emacs-lsp/dap-mode/tree/master/docs/page).

Then the final result will appear [here](https://emacs-lsp.github.io/dap-mode/)

We didn't create a Makefile for running it easily local like we [did for `lsp-mode`](https://github.com/emacs-lsp/lsp-mode/blob/master/Makefile#L72-L75), but feel free to create that too :), then you can just run `make local-webpage` to test changes locally

**Comment by nbfalcon** [comment]
I made a new PR like you said.
